enhancements/network/on-prem-dual-stack-vips.md

```diff
@@ -0,0 +1,421 @@
+validations to ensure that the VIP choices are sane - i.e. right now in a dual
+stack cluster the primary VIP must be ipv4, both VIPs should not be the same
+IP version, etc. Some amount of new validations are going to be required
+regardless of what format we pick though.
+
+
+### openshift/api
+
+We will need to make changes to the
+[baremetal platformStatus](https://github.com/openshift/api/blob/354aa98a475c1fcd60b41aaee735da95d7318b42/config/v1/0000_10_config-operator_01_infrastructure.crd.yaml#L311)
+to add new fields for the new VIPs.
+
+The new structure of platformStatus would look like this:
+
+```yaml
+baremetal:
+  description: BareMetal contains settings specific to the BareMetal platform.
+  type: object
+  properties:
+    apiServerInternalIP:
+      description: apiServerInternalIP is an IP address...
+      type: string
+    apiServerInternalIPSecondary:
+      description: apiServerInternalIPSecondary is an additional IP address...
+      type: string
+    ingressIP:
+      description: ingressIP is an external IP...
+      type: string
+    ingressIPSecondary:
+      description: ingressIPv4 is an additional external IP...
+      type: string
+    nodeDNSIP:
+    ...
+```
+
+This has the benefit of not requiring any deprecations. It should also
+provide automatic backward compatibility with existing configuration -
+in existing clusters the VIP in the original field will by definition
+be the primary, whether v4 or v6. Any other proposed structure for these
+values would require some amount of migration for existing clusters. This
+one does not.
+
+While this does not explicitly state which address is v4 and which is v6,
+that shouldn't matter. All we need to do is make sure the ip versions for
+things like unicast keepalived match the associated VIP. It doesn't matter
+to the consuming code which version that is.
+
+### machine-config-operator
+
+We will need to wire in the new VIPs to the manifests and configuration
+templates. The primary VIP will always be configured, the secondary only if
+one was specified.
+
+I believe haproxy already listens on both ipv4 and ipv6 so there should be no
+changes needed in that configuration. We'll need to verify that as part of
+this implementation though.
+
+We could also modify the apiserver configuration so it listens on both v4 and
+v6, but since there's currently no way to set up the `kubernetes.default`
+Service as dual stack this doesn't accomplish much. Traffic to the
+secondary IP version will have to come in via the VIP, which already has
+haproxy to handle both versions.
+
+### baremetal-runtimecfg
+
+There are a number of places in runtimecfg where a VIP is used to determine
+which ip version to use (for example, when deciding whether to use the ipv4
+or ipv6 addresses in the unicast_peer list). This logic will need to be
+updated to handle the fact that there may be both v4 and v6 VIPs.
+
+Additionally, the new VIPs will need to be wired in to the rendering code.
+
```

Comment:
Lets add a section here how will it affect assisted installer / ZTP. Once of the main goals of this work is to allow ZTP deployments of similar topologies

Reply:
The main goal of this enhancement is to cover hub/spoke scenario AFAIK. Im not saying it should be covered by this enhancement, but it should be considered.